In [ ]:
import re
from collections import Counter
import random
import numpy as np

In [ ]:
from google.colab import drive 
drive.mount('/content/gdrive')

Mounted at /content/gdrive


# Загрузка данных

In [ ]:
def read_file(address):
  my_file = open(address, "r")

  data = my_file.read()
  data_into_list = data.replace('\n', ' ')
  data_into_list = re.split("! |\. |\? ", data_into_list)
  
  for i in range(len(data_into_list)):
    data_into_list[i] = re.sub(r'[^\w\s]','',data_into_list[i]).lower()

  my_file.close()
  return data_into_list

In [ ]:
AnnaKarenina = read_file('gdrive/My Drive/Advanced_ML/AnnaKarenina.txt')
AnnaKarenina[:3]

['annotation   анна каренина один из самых знаменитых романов льва толстого начинается ставшей афоризмом фразой все счастливые семьи похожи друг на друга каждая несчастливая семья несчастлива посвоему',
 'это книга о вечных ценностях о любви о вере о семье о человеческом достоинстве',
 '           лев толстойроман широкого дыхания  часть перваяi   лев толстой  анна каренина     роман широкого дыхания     анна каренина поразила современников вседневностью содержания']

In [ ]:
len(AnnaKarenina)

21623

In [ ]:
WarAndPeaceEng = read_file('gdrive/My Drive/Advanced_ML/WarAndPeaceEng.txt')
len(WarAndPeaceEng)

27265

In [ ]:
WarAndPeace = read_file('gdrive/My Drive/Advanced_ML/WarAndPeace.txt')
len(WarAndPeace)

7823

# 1. Реализуйте базовый частотный метод по Шерлоку Холмсу:

In [ ]:
#AnnaKarenina
counts_AK = Counter(' '.join(AnnaKarenina))
for letter in list(counts_AK):
  counts_AK[letter] /= len(' '.join(AnnaKarenina))
print(counts_AK.most_common())

[(' ', 0.17811158425959717), ('о', 0.09392803914685217), ('е', 0.07151205931018768), ('а', 0.06772622881892615), ('н', 0.05675796189763452), ('и', 0.054291330818314255), ('т', 0.04895033714480368), ('с', 0.043447407550493644), ('л', 0.04101258531275899), ('в', 0.03849563842947604), ('р', 0.03255432516385891), ('к', 0.028026481149791307), ('д', 0.02407755805258175), ('м', 0.02343848950799819), ('у', 0.02205162401486572), ('п', 0.019717433385056438), ('я', 0.017607061331494974), ('ь', 0.01610857572152574), ('ы', 0.015160672242274645), ('г', 0.014859355761072803), ('б', 0.01429547175114613), ('ч', 0.013798097137262091), ('з', 0.013371858660014956), ('ж', 0.009265048040025934), ('й', 0.008595327339005332), ('ш', 0.006979438186456489), ('х', 0.006353671521081454), ('ю', 0.005096354764588547), ('э', 0.002902701380330222), ('щ', 0.0023446007961463876), ('ц', 0.002309321899115078), ('ф', 0.0010300281247993875), ('1', 0.0004482154950699187), ('e', 0.0004262384772471356), ('i', 0.000425660134672

In [ ]:
#WarAndPeaceEng
counts_WAPE = Counter(' '.join(WarAndPeaceEng))
for letter in list(counts_WAPE):
  counts_WAPE[letter] /= len(' '.join(WarAndPeaceEng))
print(counts_WAPE.most_common())

[(' ', 0.18627540100854006), ('e', 0.10125648325379599), ('t', 0.07272422413757106), ('a', 0.06610726603030374), ('o', 0.0619549642122142), ('n', 0.05915849638299724), ('i', 0.055981071645274515), ('h', 0.05377209975674649), ('s', 0.052322471992761174), ('r', 0.047676789220654034), ('d', 0.03799611526741023), ('l', 0.031005588117485054), ('u', 0.021018157125773276), ('m', 0.019801407741776744), ('c', 0.019794019875928822), ('w', 0.019017972750338636), ('f', 0.01763322972119158), ('g', 0.01648650445697098), ('y', 0.014860852758870016), ('p', 0.014625725897971002), ('b', 0.01113255019813935), ('v', 0.008700336276380704), ('k', 0.006562673353863035), ('x', 0.0014081914729252383), ('j', 0.0008267985518498092), ('z', 0.0007670532019492402), ('q', 0.000748422931550138), ('1', 0.0001259149309732421), ('8', 6.199383081080543e-05), ('0', 5.74968689903325e-05), ('2', 4.7218099114965795e-05), ('3', 1.9593905074917776e-05), ('6', 1.830905884049694e-05), ('5', 1.766663572328652e-05), ('7', 1.284846

In [ ]:
#WarAndPeace
counts_WAP = Counter(' '.join(WarAndPeace))
for letter in list(counts_WAP):
  counts_WAP[letter] /= len(' '.join(WarAndPeace))
print(counts_WAP.most_common())

[(' ', 0.17137623630767793), ('о', 0.08951517603735909), ('а', 0.06603719841834417), ('е', 0.06210789089671472), ('и', 0.05234889329373838), ('н', 0.05129864343944411), ('т', 0.044725452418130904), ('с', 0.041086826010555084), ('л', 0.03984376255296897), ('в', 0.036260643091795344), ('р', 0.035889622976370106), ('к', 0.028232585791098146), ('д', 0.023936640281391003), ('м', 0.023283703306607225), ('у', 0.0225737986763054), ('п', 0.02022643912713866), ('я', 0.01822526763842775), ('г', 0.016326345787826158), ('ь', 0.015334524298165788), ('ы', 0.014947436382466232), ('з', 0.014025728930366535), ('б', 0.013599201868539102), ('ч', 0.01073475129236239), ('й', 0.009071003609412225), ('ж', 0.007975471772526692), ('ш', 0.0074350093996631604), ('х', 0.0067192619328979445), ('e', 0.0064986948564049905), ('ю', 0.00510517835988659), ('ц', 0.0031828851634314393), ('n', 0.0030178250333406857), ('s', 0.0030061393604139067), ('i', 0.00290388972230459), ('a', 0.0027563581016040047), ('r', 0.002706693991

In [ ]:
def encrypt_frequency_method(counts, text):
  list_a = list(counts)
  list_b = list(counts)
  random.shuffle(list_b)
  mapping = zip(list_a, list_b)
  mapping = dict(mapping)
  decoder = ''.join(mapping[i] for i in text)
  return decoder

In [ ]:
encrypt_AK = encrypt_frequency_method(counts_AK, ' '.join(AnnaKarenina[:4]))
encrypt_AK

'héénwhwênéшшшáыыáшâázäылыáш7aлышлbш6áдgбшbыáдäылмgбшz7дáы7зшйsзáшм7й6м7u7шыáьлыáäм6pш6мáзтäршá37zлbд7дш3záb7ршз6äш6ьá6мйлзgäш6äдsлша7б7влшazduшыáшazduáшâáвaápшыä6ьá6мйлзápш6äдspшыä6ьá6мйлзáша76з7äдdш1м7шâылuáш7шзäьыgбшфäыы76мpбш7шй óзлш7шзäzäш7ш6äдsäш7шьäй7зäьä6â7дшa76м7лы6мзäшшшшшшшшшшшшйäзшм7й6м7рz7дáыштлz7â7u7шagбáылpшшьá6мsшаäzзápêшшшйäзшм7й6м7ршшáыыáшâázäылыáшшшшшz7дáыштлz7â7u7шagбáылpшшшшшáыыáшâázäылыáша7zábлйáш67зzäдäыылâ7зшз6äaыäзы76мs ш67aäzвáылpшыä7ógьáрыápш6з7ó7aáшzá6â7зáыы76мsша7зä6мз7зáылpшdaлзлмäйsы7ш67ьäмáйл6sшзш1м7дшz7дáыäш6шфäйsы76мs шбda7вä6мзäыы7u7шзbuйpaáшáзм7záшыáшвлbыs'

In [ ]:
encrypt_WAPE = encrypt_frequency_method(counts_WAPE, ' '.join(WarAndPeaceEng[:4]))
encrypt_WAPE

'nféu5b8yéenucänérhébcuéh88vu8 ul4bu4rku5é4eéuhwu0é8un80àn8wuunf7àuéh88vu7àu 8bunféuäàéu8 u4rw8réu4rwlfébéu4nur8ue8ànu4rkul7nfu40j8ànur8ubéànb7en78ràulf4nà8é2ébuuw8äuj4wue85wu7nuc72éu7nu4l4wu8bubéäàéu7nuärkébunféunébjàu8 unféu5b8yéenucänérhébcu07eéràéu7re0äkékul7nfunf7àuéh88vu8bu8r07réu4nulllcänérhébc8bcuuun7n0éul4bu4rku5é4eéuu4änf8bu0é8un80àn8wuunb4rà04n8bàu08ä7àéu4rku4w0jébuj4äkéuu58àn7rcuk4néuy4rä4bwua1u611suéh88vu6d11uu04ànuä5k4nékuj4befua9u61aêuu04rcä4céuérc07àfuuuuàn4bnu8 unf7àu5b8yéenucänérhébcuéh88vul4bu4rku5é4eéuuu4ru4r8rwj8äàu280ärnéébu4rkuk427kul7kcébuuuuuul4bu4rku5é4eéuuhwu0é8un80àn8wn80àn87uue8rnérnàuuh88vu8réuam19uuef45nébu7uuef45nébu77uuef45nébu777uuef45nébu72uuef45nébu2uuef45nébu27uuef45nébu277uuef45nébu2777uuef45nébu73uuef45nébu3uuef45nébu37uuef45nébu377uuef45nébu3777uuef45nébu372uuef45nébu32uuef45nébu327uuef45nébu3277uuef45nébu32777uuef45nébu373uuef45nébu33uuef45nébu337uuef45nébu3377uuef45nébu33777uuef45nébu3372uuef45nébu332uuef45nébu3327uuef45nébu33277uuef45nébu33277

In [ ]:
encrypt_WAP = encrypt_frequency_method(counts_WAP, ' '.join(WarAndPeace[:4]))
encrypt_WAP

'лppнzлzhнpâââбjcà7уâоâiоéââôуisàâоmjxô67sàâéciу7âyujуâ7одcyуxjоdуâ6cyô6cвcâдудâ7одудcxâщéъвcxâüécоmjxщx7оxâüоôу6xy3âc6éуеуx6âвyъäо7ъâxвcâiоéccкъкx7о3âоâюоycôcюооб\t6уâд7овуâоmâéуmé3щуâjxd7sшâüc6ciъâd6câc7уâcäcâjôxiââcâеоm7оâоâôixé6оâcâyfäjоâоâdxô6оâcâiъеxô6jxâоâвxécоmixâcâôyуjxâоâücщjовxâcâjcà7xâоâiоéxбüxéjsàâ6ciâm7удciо6âôâjsôwоiâcäкxô6jciâécôôооâиhиâjxдуâücдуmу7sâjmуоicc67cwx7о3âixещъâécщо6xy3iоâоâщx6uiоâjâôxiuxâécô6cjsшâôjу6cjô6jcâъâäcyдc7ôдошâо76éовоâъâäxmъшcjsшâjxdxéуâjâôуyc7xâюéxàyо7sâуüwxéxéâäуysâjâicôдjxâоâüx6xéäъéвxââââââââââââyxjâ7одcyуxjоdâ6cyô6càdуô6uâüxéjу3ââdуô6uâqzcéу3ââdуô6uâ6éx6u3ââââââââââââââyxjâ7одcyуxjоdâ6cyô6càââjcà7уâоâiоéââ6ciâзâââââdуô6uâüxéjу3ââââââhââââxйâqhпpâтнpâэthpîпâгçpпыâпzâ1аîaапыâpпâынpzâэ1аыâaапâцпыâлэлpлгпыâцпыâücixô6u3âцпâ1лâ0лтh11пâqанpлэлtzп'

In [ ]:
def decryption_frequency_method(encrypt, mydict):
  counts_dec = Counter(encrypt)
  for letter in list(counts_dec):
    counts_dec[letter] /= len(encrypt)
  
  de_counts_dec =[]
  input_list = list(mydict.values())
  for i in encrypt:
    difference = lambda input_list : abs(input_list - counts_dec[i]) 
    res = min(input_list, key=difference)
    letter = list(mydict.keys())[input_list.index(res)]
    de_counts_dec.append(letter)
  return ''.join(de_counts_dec)

In [ ]:
decryption_AK = decryption_frequency_method(encrypt_AK, counts_AK)
decryption_AK

'эююээээээю   оеео ьораетео оуте тз нодзз зеодаетсзз родоеот кпто сокнсозо еоьтеоаснп нсотюай оэортздод эрозой тна ньонскттза надпт йозойт урйз ео урйзо ьойуоп еаньонскттоп надпп еаньонсктто йонтоадй эсо ьетзо о таьезз эаееонспз о кюютт о тара о надпа о ьакотаьаньод уонсотенста            кат сокнсойродое ютроьозо уззоетп  ьонсп йартопэ   кат сокнсой  оеео ьораетео     родое ютроьозо уззоетп     оеео ьораетео йорозтко нотрадаеетьот тнауеатеонспю ноуарйоетп еаоюзьойеоп нтоюоуо роньотоееонсп йотанстотоетп йутттсакпео ноьасоктнп т эсод родоеа н эакпеонспю зйуойанстаееозо тззкпуо отсоро ео йтзеп'

In [ ]:
decryption_WAPE = decryption_frequency_method(encrypt_WAPE, counts_WAPE)
decryption_WAPE

'tat otcjtat xkttbkttx tkccx cx xtt tbk ottat kk ktc tckvtck  taev tkccx ev xct tat kvt cx tbkcbt tbkxattt tt bc acvt tbk xeta tkxcvt bc ttvtteatecbv xattvctltt  kck xtk acok et xelt et txtk ct ttkvt et kbktt tat tttxv cx tat otcjtat xkttbkttx keatbvt ebakkktk xeta taev tkccx ct cbkebt tt xxxxkttbkttxctx   tetkt xtt tbk ottat  tktact ktc tckvtck  tttbvkttctv kckevt tbk tkkxtt xtkkt  ocvtebx kttt jtbkttk kx xxx1 tkccx xqxx  ktvt koktttk xttaa kj xxkq  ktbxktxt tbxkeva    vtttt cx taev otcjtat xkttbkttx tkccx xtt tbk ottat   tb tbcbkxckv lckkbtttt tbk ktlek xekxtt      xtt tbk ottat  kk ktc tckvtcktckvtce  acbttbtv  kccx cbt kxxj  aatottt e  aatottt ee  aatottt eee  aatottt el  aatottt l  aatottt le  aatottt lee  aatottt leee  aatottt eh  aatottt h  aatottt he  aatottt hee  aatottt heee  aatottt hel  aatottt hl  aatottt hle  aatottt hlee  aatottt hleee  aatottt heh  aatottt hh  aatottt hhe  aatottt hhee  aatottt hheee  aatottt hhel  aatottt hhl  aatottt hhle  aatottt hhlee  aatottt hhlee

In [ ]:
decryption_WAP = decryption_frequency_method(encrypt_WAP, counts_WAP)
decryption_WAP

'чззeшчшшeз   цтачкт а кар  вткбч абтавткбч рактк дчтт каяадтатазт тадвтача ятя каятяаа йрзчаа зраабтайакаа завттадб атртютат чдзчакз ача караацзцакаб а цадавацаацdтт якачт аб ртбрбйт тазкбe затакз зта акт ача твак  а юабка а вкарта а дdчта а завта а кзюавтта а чараабка а вдтта а зайтача а тачка а карацзартбч так бктяакат в тбвцак ачцавттак равваа uшu таят заятбткб тбтакааткацакаб каюйз райатадбка а йатчка т вакча равтатбe втттатвтта з чадяаквяаe актрача з чабзeатбe тазарт т втдака црачдакб тзцарар чтдб т кавята а затарчзрча            дат каяадтатаз тадвтачзтвтч зарттб  зтвтч цшартб  зтвтч тратчб              дат каяадтатаз тадвтач  тачкт а кар  так d     зтвтч зарттб      ш    аd цшгз ueз юuшзuг udзгй гш eeuueгй зг йeзш юeeй ueг цгй чючзчuгй цгй закавтчб цг eч dчuшeeг цeeзчючuшг'

In [ ]:
def probability_decryption(decryption, text):
  count = 0
  for i in range(min(len(text), len(decryption))):
    if decryption[i] == text[i]:
      count += 1
  return count / len(text)

In [ ]:
#AnnaKarenina
probability_AK = probability_decryption(decryption_AK, ' '.join(AnnaKarenina[:4]))
probability_AK

0.3232830820770519

In [ ]:
#WarAndPeaceEng
probability_WAPE = probability_decryption(decryption_WAPE, ' '.join(WarAndPeaceEng[:4]))
probability_WAPE

0.3055601659751037

In [ ]:
#WarAndPeace
probability_WAP = probability_decryption(decryption_WAP, ' '.join(WarAndPeace[:4]))
probability_WAP

0.28607594936708863

### __Качество предсказания букв на базовом частотном методе по Шерлоку Холмсу около 30 процентов__: AnnaKarenina - 0.3232830820770519; WarAndPeaceEng - 0.3055601659751037; WarAndPeace - 0.28607594936708863

#2. Реализуйте частотный метод биграмм по Шерлоку Холмсу:

In [ ]:
def frequency_bigrams(text):
  list_a = []
  list_b = []
  for i in range(len(text) - 1):
    if text[i: i+2] not in list_a:
      list_a.append(text[i: i+2])
      list_b.append(text.count(text[i: i+2]) / (len(text) / 2))
  return dict(zip(list_a, list_b))

In [ ]:
frequency_bigrams_AK = frequency_bigrams(' '.join(AnnaKarenina))

In [ ]:
frequency_bigrams_WAPE = frequency_bigrams(' '.join(WarAndPeaceEng))

In [ ]:
frequency_bigrams_WAP = frequency_bigrams(' '.join(WarAndPeace))

In [ ]:
def encrypt_frequency_method_bigrams(counts, text):
  list_a = list(counts)
  list_b = list(counts)
  random.shuffle(list_b)
  mapping = zip(list_a, list_b)
  mapping = dict(mapping)
  decoder = ''
  for i in range(0, len(text) - 1, 2):
    decoder += ''.join(mapping[text[i:i+2]])
  return decoder

In [ ]:
#AnnaKarenina
encrypt_AK_bigrams = encrypt_frequency_method_bigrams(frequency_bigrams_AK, ' '.join(AnnaKarenina[:4]))
encrypt_AK_bigrams

'opллdu97лбяхwnэсимывафst6 смтпеnймеиn хтрю6 чрstфьcie7сыюфibтуg8вът9рццуcé6 g36  2ьоеичмsèаlwnâcnnх тзеу  msвылирlétнрбитрсреи2йsaобн1иэрз00днымим00днимывсe 1ымбцünрц 3g8впгоdoвпкнétнрбитриммоrt19s1уф 6лжstдосми9зшююciемэско9 cicéр1дирзcéеяафсмеи2йruсмяк58a1зшбцt1х8ghрца льткрlяхяхяхяхяхиафювът9рцбъe7сыеnбнe7t1rbн3хтmme6яхünрци3ъ щр 1реяхttib 6ё  6выwnэсимывафst6 яхяхhüтзmmп myушцуcéлзапstвпяхяхmm6 лж 8 м61иммо  imдюеиa1афчрэслгa1и9гов2фююфрц42еи47чу вstвпкндзвбез6 впrtдз47им  дхa1mmюфрци3моеярцsgg8stвпl2трff58зюcéрыпбчм 38ми9уф 6х8e7сыкнеиio58зюкоаэéqдгghóмрцеяэсцуcé63akabимяч 6  ымимdicr'

In [ ]:
#WarAndPeaceEng
encrypt_WAPE_bigrams = encrypt_frequency_method_bigrams(frequency_bigrams_WAPE, ' '.join(WarAndPeaceEng[:4]))
encrypt_WAPE_bigrams

'sbegsomm 7fgbdti 6ypharjwgxy31mioxgsp é bc21hkrqpkncpuirpurqno6suqrjwgxyigdbmpnoihmhbuoqfkttste8gs xhwypegpt80ncldktgsp mi0 tef441kt80nc1akttyyk uywmi56ohfiedv aq byourinenrq0 atpaeg0 gsgfrqmpkylybulbfgyggnv sbegtiewuq31noihé lwfnykatzvhrpwlnnwuphrbulbroyjgnksjosbno6suqrjwgxympoqxuwsegptmiaobdti 6yp89ln87no0 pkmioxgsp é bc2187eksbmpnw2 no3rkth 87 3ttpsptmpuqvbeibugsp gs 8llv e7eeegé ysnjosagpteg2h joxrqwtxcbvskrjwgxyvsbv87zaktmhugpti yooxffeygdgbna87zaos fotcloswlf38787ktoxfg31no6suqsomm 7fgbdti 6ypharjwgxygfv ttks mvreg87ttgse6 x41eà 43rygtiypgsp aghgwimiwiotv 8787mioxgsp é bc2187spnw2 no3rkth puirpuuwindttihvuqhkwgxydtegdd9t87ffhdtiv uwin56iwyplbuwin56iwyplb k87ffhdtiv pa87ffhdtiv ssin56iwyp 4uwin56iwyp 4 k87ffhdtiv df k87ffhdtiv 4587ffhdtiv euin56iwypfpuwin56iwypfp k87ffhdtiv su k87ffhdtiv sussin56iwypfpssin56iwypfpdf87ffhdtiv on k87ffhdtiv on kuwin56iwypfp4587ffhdtiv uy87ffhdtiv uyuwin56iwypfpsuuwin56iwypfpsu k87ffhdtiv uypa87ffhdtiv uyssin56iwypfponuwin56iwypfpon k87ffhdtiv uydf 

In [ ]:
#WarAndPeace
encrypt_WAP_bigrams = encrypt_frequency_method_bigrams(frequency_bigrams_WAP, ' '.join(WarAndPeace[:4]))
encrypt_WAP_bigrams

'допфzwаnши3юlâznéuуйucбоnbтдюхмхdmмк сplмхошнюр ежнмуйzaéaпеанииуйсуйрсухеелдстрыздсэцlrцохе 8orжьмктф нpoийрхюч цжунпьи жnsodзллюкжт хесжснêmсэитzaжуucяжгчônяж гikуйдэènуйimошlé тхиаюavpêhtбьсуgèнысуыкebыкénаюoehlыкжнimzadmтдйнхвucyeгннтucucущилucyegèлвиляюdmodgoжьтю 8yeас 3 8ucбьtgèn 8yeznéu 8ucбоvегдgoяеptсуhlzmдснюktтдаюдлjuhlвгитилznhlяцекiéкюсhаючруйбьtôlypêаюlyдфêmplш  нcлсжтзюгошишkt цааucucеюwoбоаютдloжюошмдсуяеhtuyючлчилznьбnkиcéa6 тмhtvurvènuclàdhsbkiяеhtяюущьиаютды  z 8фоl ivpêоов gogonkы лётаspжеяюdmийnsgoзюрж 83ю3ю3ю3ю3юежантрызиcик80apсуйрсуоэктwoийgoрфжуныктwonlяйьижуныктwoзяvеwoжу3ю3ю3ю3ю3ю3юежантрызиcик80apсуйрсуptаюбзebdmсжсн3юсуhlsu3ю3юбкилоеpôжмтш3ю3ю3юnr3ют hatyуцэршизбv кнoгamoтэлсмчтioohбяtnoгьсübзбceэлohoгулэлнодоп бяmoбяийню сwoжуулчтbивк ффмoгуэшино хrs'

In [ ]:
def decryption_frequency_method_bigrams(encrypt, mydict):
  counts_dec = frequency_bigrams(encrypt)
  
  de_counts_dec =[]
  input_list = list(mydict.values())
  for i in range(0, len(encrypt) - 1, 2):
    difference = lambda input_list : abs(input_list - counts_dec[encrypt[i: i+2]]) 
    res = min(input_list, key=difference)
    letter = list(mydict.keys())[input_list.index(res)]
    de_counts_dec.append(letter)
  return ''.join(de_counts_dec)

In [ ]:
#encrypt_AK_bigrams
decryption_AK_bigrams = decryption_frequency_method_bigrams(encrypt_AK_bigrams, frequency_bigrams_AK)
decryption_AK_bigrams

'ияияияияияе ерго сергоя  ороиявсияя иявсия овся ияерроерервсияервсвс сер   оия оияияя всияияерияияиявсия  иявсиявсвсвсвсроияя всияияияиявсвсвсер свсвс сериявсервсвс свсер  всия  ервсвсвсро сервсияиявсговся ияроервсвсерияговсияер  ияиявс  рогороя всияроияерровсвсервсвс сияияиявсе е е е е иявсвсвс сияроервсвсроерияиявсроияе вс свсияиявсвсе иявсгоияговсерго сергоя  ое е иявсроияияияер  ияияя   е е ро овсияияия сер  ияияя роговсгоияроервсиявсер сияя всиявся   ервсияия о  всвсвс с  ияророер свсерро сияеря   ияроияервс  ияиявсвсияервсговсроереря ияервсвсияияиявсия срогоер  ияияия сияго  ер сияия'

In [ ]:
#AnnaKarenina
probability_AK_bigrams = probability_decryption(decryption_AK_bigrams, ' '.join(AnnaKarenina[:4]))
probability_AK_bigrams

0.10217755443886097

In [ ]:
#encrypt_WAPE_bigrams
decryption_WAPE_bigrams = decryption_frequency_method_bigrams(encrypt_WAPE_bigrams, frequency_bigrams_WAPE)
decryption_WAPE_bigrams

'ctooknbiysapkne ys tbiciw weknmictw  iciysknwectknctelknelctw ciniciw weysbituw tubimiysbifrknysw bixi tooctcictximiw  imicictbibimicictysmibiysg ximi txixitu ttumiapbi txictcibieloociw yscttuxiximi bapknci tctooe xiniknw tucixixiysbibicixibiysbicimi bxibicituknctw ciniciw wetuysxibiooctmibikne ys txibie w ciknmictw  iciyskne xicttuysbiw ysmiyse xifrnccttuniysximiw  iw cibi txixioocimiknctcictooxixictctxiysbixiciw webibie ysmibiknctknapct twectyskne ysctbiknysctbixie e mictapknw ciniknbiysapkne ys tbiciw weys tfrtuysxiooe frw ysbibixi cyskne  tw  icixit mit kn te e mictw  iciyskne frysbiw ysmiyselknels  t te biniwew we toocixie  t te  ts  t t t t bs  t t t t bt e  t te  tele  t te  tar t t t t cs  t t t t ct e  t te  to t e  t te  twee  t te  th  t t t t hs  t t t t ht e  t te  t it e  t te  t iar t t t t har t t t t ho e  t te  test e  t te  test s  t t t t hwee  t te  tite  t te  tits  t t t t h is  t t t t h it e  t te  titele  t te  titar t t t t hess  t t t t hest e  t te  tito t

In [ ]:
#decryption_WAPE_bigrams
probability_WAPE_bigrams = probability_decryption(decryption_WAPE_bigrams, ' '.join(WarAndPeaceEng[:4]))
probability_WAPE_bigrams

0.1646473029045643

In [ ]:
#encrypt_WAP_bigrams
decryption_WAP_bigrams = decryption_frequency_method_bigrams(encrypt_WAP_bigrams, frequency_bigrams_WAP)
decryption_WAP_bigrams

'мелслслсзао лсзаме о в злсерлсмеермемемемеза злс злс о змелс злс о вза в злсза з ззалслслс з  лсмемелсмелсзалс змеколс злсмемелслсмеме з з змемеме зко вмелслсмелслс олс з омезалсмелс илс зза з в з з вермеерлс илсерермеме зерермелс взалс з в вмеер вза злсер зермекомелс  залслс   в злс з  зазаме   в змелско зме верлсза змеер илслсерлсмеерзаерлслслслслс илс о злсме з имелсмемелсмелс злслсзалсмемелс в влско з иерлслсзалс в ззалс змеерзалсме змелслсзалслс з влслслслс зза зме з иермелс  лслслс злслскокомемелслслслс зерзамекомелс  о о о о о  з з з з змемеме вза влсзакозаколско ззаколслс зко ззаколсмекокоо о о о о о  з з з з змемеме вза вме илсмеер з зо  верлсо о лсерлслслслсо о о лсо мелслслслсзамелслсзалслс злсмелсме злсзамелсмелс змезаме змемелс з з зза змекокомемелслслслсзалсзамемелс'

In [ ]:
#decryption_WAP_bigrams
probability_WAP_bigrams = probability_decryption(decryption_WAP_bigrams, ' '.join(WarAndPeace[:4]))
probability_WAP_bigrams

0.08734177215189873

### __Качество предсказания букв на частотном методе биграмм по Шерлоку Холмсу около 10 процентов__: AnnaKarenina - 0.10217755443886097; WarAndPeaceEng - 0.1646473029045643; WarAndPeace - 0.08734177215189873

#3. Но и это ещё не всё: биграммы скорее всего тоже далеко не всегда работают.

###Основная часть задания — в том, как можно их улучшить:
###○ предложите метод обучения перестановки символов в этом задании, основанный на MCMC-сэмплировании, но по-прежнему работающий на основе статистики биграмм;
###○ реализуйте и протестируйте его, убедитесь, что результаты улучшились.


In [ ]:
def read_file_leng(address, leng):
  my_file = open(address, "r")

  data = my_file.read()
  data_into_list = data.replace('\n', ' ')
  data_into_list = re.split("! |\. |\? ", data_into_list)
  
  for i in range(len(data_into_list)):
    data_into_list[i] = re.sub(leng,'',data_into_list[i]).lower()

  my_file.close()
  return data_into_list

In [ ]:
AnnaKarenina_rus = read_file_leng('gdrive/My Drive/Advanced_ML/AnnaKarenina.txt', r'[^а-яА-Я\ ]*')
AnnaKarenina_rus[:3]

['   анна каренина один из самых знаменитых романов льва толстого начинается ставшей афоризмом фразой все счастливые семьи похожи друг на друга каждая несчастливая семья несчастлива посвоему',
 'это книга о вечных ценностях о любви о вере о семье о человеческом достоинстве',
 '           лев толстойроман широкого дыхания  часть первая   лев толстой  анна каренина     роман широкого дыхания     анна каренина поразила современников вседневностью содержания']

In [ ]:
WarAndPeaceEng_eng = read_file_leng('gdrive/My Drive/Advanced_ML/WarAndPeaceEng.txt', r'[^a-zA-Z\ ]*')
WarAndPeace_rus = read_file_leng('gdrive/My Drive/Advanced_ML/WarAndPeace.txt', r'[^а-яА-Я\ ]*')

In [ ]:
frequency_bigrams_AK_rus = frequency_bigrams(' '.join(AnnaKarenina_rus))
frequency_bigrams_WAP_rus = frequency_bigrams(' '.join(WarAndPeace_rus))
frequency_bigrams_WAPE_eng = frequency_bigrams(' '.join(WarAndPeaceEng_eng))

In [ ]:
counts_AK = Counter(' '.join(AnnaKarenina_rus))
for letter in list(counts_AK):
  counts_AK[letter] /= len(' '.join(AnnaKarenina_rus))

counts_WAP = Counter(' '.join(WarAndPeace_rus))
for letter in list(counts_WAP):
  counts_WAP[letter] /= len(' '.join(WarAndPeace_rus))

counts_WAPE = Counter(' '.join(WarAndPeaceEng_eng))
for letter in list(counts_WAPE):
  counts_WAPE[letter] /= len(' '.join(WarAndPeaceEng_eng))

In [ ]:
encrypt_AK_bigrams_rus = encrypt_frequency_method(counts_AK, ' '.join(AnnaKarenina_rus[:4]))

encrypt_WAP_bigrams_rus = encrypt_frequency_method(counts_WAP, ' '.join(WarAndPeace_rus[:4]))

encrypt_WAPE_bigrams_eng = encrypt_frequency_method(counts_WAPE, ' '.join(WarAndPeaceEng_eng[10:15]))

encrypt_AK_bigrams_rus

'фффсввсфжсрьвцвсфеуцвфцпфэсздщфпвсзьвцхдщфрезсве фмш сфхемэхеоефвснцвсьхэыфэхс йьбфсюерцпзезфюрспебф эьфэнсэхмц дьфэьзшцфкещетцфурлофвсфурлосфжстусыфвьэнсэхмц сыфэьзшыфвьэнсэхмц сфкеэ еьзлфъхефжвцосфеф ьнвдщфаьввеэхыщфефмич цфеф ьрьфефэьзшьфефньме ьньэжезфуеэхецвэх ьффффффффффффмь фхемэхебрезсвфйцрежеоефудщсвцыффнсэхшфкьр сыфффмь фхемэхебффсввсфжсрьвцвсфффффрезсвфйцрежеоефудщсвцыфффффсввсфжсрьвцвсфкерспцмсфэе рьзьввцже ф эьувь веэхшифэеуьртсвцыфвьечднсбвсыфэ ечеусфрсэже сввеэхшфке ьэх е свцыфлуц цхьмшвефэеньхсмцэшф фъхезфрезсвьфэфаьмшвеэхшифщлуетьэх ьввеоеф помыусфс херсфвсфтцпвш'

In [ ]:
' '.join(WarAndPeaceEng_eng[10:15])

'what a virulent attack replied the prince not in the least disconcerted by this reception he had just entered wearing an embroidered court uniform knee breeches and shoes and had stars on his breast and a serene expression on his flat face he spoke in that refined french in which our grandfathers not only spoke but thought and with the gentle patronizing intonation natural to a man of importance who had grown old in society and at court he went up to anna pavlovna kissed her hand presenting to her his bald scented and shining head and complacently seated himself on the sofa  first of all dear friend tell me how you are'

In [ ]:
def sum_prob(decryption, frequency_bigrams):
  sum_dec = 0
  for i in range(0, len(decryption) - 1, 2):
    try:
      sum_dec += frequency_bigrams[decryption[i: i+2]]
    except:
      sum_dec -= min(frequency_bigrams.values())
  for i in range(1, len(decryption) - 2, 2):
    try:
      sum_dec += frequency_bigrams[decryption[i: i+2]]
    except:
      sum_dec -= min(frequency_bigrams.values())
  return sum_dec

In [ ]:
def decryption_mcmc_bigrams(encrypt, frequency_bigrams_ex, frequency_ex, count_iter):
  counts_dec = Counter(encrypt)
  list_a = [i[0] for i in counts_dec.most_common()]
  list_b = [i[0] for i in frequency_ex.most_common(len(list_a))]
  mapping_old = dict(zip(list_a, list_b))
  decryption_old = ''
  for i in encrypt:
    decryption_old += mapping_old[i]

  sum_old = sum_prob(decryption_old, frequency_bigrams_ex)
  sum_best = sum_old
  mapping_best = mapping_old.copy()
  decryption_best = decryption_old
  for _ in range(count_iter):
    mapping_new = mapping_old.copy()
    ran_1 = random.choice(list(mapping_new))
    ran_2 = random.choice(list(mapping_new))
    if ran_1 != ran_2:      
      mapping_new[ran_1], mapping_new[ran_2] = mapping_new[ran_2], mapping_new[ran_1]
      decryption_new = ''
      for i in encrypt:
        decryption_new += mapping_new[i]
      sum_new = sum_prob(decryption_new, frequency_bigrams_ex)

      if sum_new >= sum_old:
        sum_old = sum_new
        mapping_old = mapping_new.copy()
        if sum_new >= sum_best: 
          sum_best = sum_new
          mapping_best = mapping_new.copy()
      else:
        if sum_new - sum_old > random.random():
          sum_old = sum_new          
          mapping_old = mapping_new.copy()
  decryption_best = ''
  for i in encrypt:
    decryption_best += mapping_best[i]    
  return decryption_best

In [ ]:
decryption_AK_bigrams_rus = decryption_mcmc_bigrams(encrypt_AK_bigrams_rus, frequency_bigrams_AK_rus, counts_AK, 1000)
print(decryption_AK_bigrams_rus)
print(probability_decryption(decryption_AK_bigrams_rus, ' '.join(AnnaKarenina_rus[:4])))

   енне перинане одан аз селшя знелинатшя роленов кьве токстомо нечанеитсу стевхий ещоразлол щрезой вси счесткавши сильа гояожа дрцм не дрцме пеждеу нисчесткавеу сильу нисчесткаве госвоилц это пнаме о вичншя бинностуя о кыюва о вири о сильи о чиковичиспол достоанстви            кив токстойролен харопомо дшяенау  честь гирвеу   кив токстой  енне перинане     ролен харопомо дшяенау     енне перинане горезаке соврилиннапов всиднивностьы содирженау ниоюшчейнеу своюоде респовенность говиствовенау цдаватикьно сочитекась в этол ролени с бикьностьы яцдожиствинномо взмкуде евторе не жазнь
0.6279863481228669


In [ ]:
#AnnaKarenina_rus
decryption_AK_bigrams_rus = decryption_mcmc_bigrams(encrypt_AK_bigrams_rus, frequency_bigrams_AK_rus, counts_AK, 100000)
print(decryption_AK_bigrams_rus)
print(probability_decryption(decryption_AK_bigrams_rus, ' '.join(AnnaKarenina_rus[:4])))

   анна каленина один их самбы хнаменитбы ломанов рьва торстого назинается ставшей ацолихмом цлахой все сзастривбе семьи поыожи длуг на длуга каждая несзастривая семья несзастрива посвоему это книга о везнбы ченностяы о рющви о веле о семье о зеровезеском достоинстве            рев торстойломан шилокого дбыания  засть пелвая   рев торстой  анна каленина     ломан шилокого дбыания     анна каленина полахира совлеменников вседневностью соделжания неощбзайная свощода ласкованность повествования удивитерьно созетарись в этом ломане с черьностью ыудожественного вхгряда автола на жихнь
0.8720136518771331


In [ ]:
#WarAndPeace_rus
decryption_WAP_bigrams_rus = decryption_mcmc_bigrams(encrypt_WAP_bigrams_rus, frequency_bigrams_WAP_rus, counts_WAP, 10000)
print(decryption_WAP_bigrams_rus)
print(probability_decryption(decryption_WAP_bigrams_rus, ' '.join(WarAndPeace_rus[:4])))

   нойла и вир  савый изнестлый ровал дьна ликодаенима тодстого как ликакое чругое произнечелие писатедя отражает гдубилу его вирооцуцелия и шидосошииэта клига из разряча немлых потову мто ола обо нсев  о жизли и сверти о дюбни и мести о вужестне и героизве о сдане и почниге о нойле и виреперный тов злаковит с нысщив обцестнов россии  нека показалы нзаивоотлощелия вежчу рочитедяви и четьви н севье ростоных снатонстно у бодколских илтриги у безухоных немера н садоле шрейдилы апщерер бады н воскне и петербурге            ден ликодаеним тодстоймасть перная  масть орая  масть третья              ден ликодаеним тодстой  нойла и вир  тов      масть перная          е              повестья    
0.8170028818443804


In [ ]:
#WarAndPeaceEng_eng
decryption_WAPE_bigrams_eng = decryption_mcmc_bigrams(encrypt_WAPE_bigrams_eng, frequency_bigrams_WAPE_eng, counts_WAPE, 10000)
print(decryption_WAPE_bigrams_eng)
print(probability_decryption(decryption_WAPE_bigrams_eng, ' '.join(WarAndPeaceEng_eng[10:15])))

what a kirulent attacv replied the prince not in the least disconcerted by this reception he had just entered wearing an embroidered court uniform vnee breeches and shoes and had stars on his breast and a serene expression on his flat face he spove in that refined french in which our grandfathers not only spove but thought and with the gentle patronizing intonation natural to a man of importance who had grown old in society and at court he went up to anna paklokna vissed her hand presenting to her his bald scented and shining head and complacently seated himself on the sofa  first of all dear friend tell me how you are
0.987220447284345


In [ ]:
# при использовании первых 4 предложений качество получается не очень, т.к. вначале идет оглавление chapter...
' '.join(WarAndPeaceEng_eng[:4])

'the project gutenberg ebook of war and peace by leo tolstoy  this ebook is for the use of anyone anywhere at no cost and with almost no restrictions whatsoever  you may copy it give it away or reuse it under the terms of the project gutenberg license included with this ebook or online at wwwgutenbergorg   title war and peace  author leo tolstoy  translators louise and aylmer maude  posting date january   ebook   last updated march    language english    start of this project gutenberg ebook war and peace   an anonymous volunteer and david widger      war and peace  by leo tolstoytolstoi  contents  book one   chapter i  chapter ii  chapter iii  chapter iv  chapter v  chapter vi  chapter vii  chapter viii  chapter ix  chapter x  chapter xi  chapter xii  chapter xiii  chapter xiv  chapter xv  chapter xvi  chapter xvii  chapter xviii  chapter xix  chapter xx  chapter xxi  chapter xxii  chapter xxiii  chapter xxiv  chapter xxv  chapter xxvi  chapter xxvii  chapter xxviii  book two   chapte

In [ ]:
encrypt_WAPE_bigrams_eng_chapter = encrypt_frequency_method(counts_WAPE, ' '.join(WarAndPeaceEng_eng[:4]))

In [ ]:
decryption_WAPE_bigrams_eng_chapter = decryption_mcmc_bigrams(encrypt_WAPE_bigrams_eng_chapter, frequency_bigrams_WAPE_eng, counts_WAPE, 10000)
print(decryption_WAPE_bigrams_eng_chapter)
print(probability_decryption(decryption_WAPE_bigrams_eng_chapter, ' '.join(WarAndPeaceEng_eng[:4])))

sri unlxias vksimwinv iwlly lp bon omg uioai wf cil slcdslf  sred iwlly ed pln sri kdi lp omflmi omfbrini os ml alds omg besr ocjlds ml nidsneaselmd brosdlihin  flk jof aluf es vehi es obof ln nikdi es kmgin sri sinjd lp sri unlxias vksimwinv ceaimdi emackgig besr sred iwlly ln lmcemi os bbbvksimwinvlnv   sesci bon omg uioai  oksrln cil slcdslf  snomdcoslnd clkedi omg ofcjin jokgi  uldsemv gosi xomkonf   iwlly   cods kugosig jonar    comvkovi imvcedr    dsons lp sred unlxias vksimwinv iwlly bon omg uioai   om omlmfjlkd hlckmsiin omg goheg begvin      bon omg uioai  wf cil slcdslfslcdsle  almsimsd  wlly lmi   arousin e  arousin ee  arousin eee  arousin eh  arousin h  arousin he  arousin hee  arousin heee  arousin et  arousin t  arousin te  arousin tee  arousin teee  arousin teh  arousin th  arousin the  arousin thee  arousin theee  arousin tet  arousin tt  arousin tte  arousin ttee  arousin tteee  arousin tteh  arousin tth  arousin tthe  arousin tthee  arousin ttheee  wlly sbl   arousin

### __Качество предсказания букв на методе основанном на MCMC-сэмплировании биграмм около 85 процентов__: AnnaKarenina - 0.8720136518771331; WarAndPeace - 0.8170028818443804; WarAndPeaceEng(c 10 по 14 предложение) - 0.987220447284345б WarAndPeaceEng(при использовании первых 4 предложений качество получается не очень, т.к. вначале идет оглавление chapter...) - 0.23326024624725925

#4. Расшифруйте сообщение:

In [ ]:
text ='←⇠⇒↟↹↷⇊↹↷↟↤↟↨←↹↝⇛⇯↳⇴⇒⇈↝⇊↾↹↟⇒↟↹⇷⇛⇞↨↟↹↝⇛⇯↳⇴⇒⇈↝⇊↾↹↨←⇌⇠↨↹⇙↹⇸↨⇛↙⇛↹⇠⇛⇛↲⇆←↝↟↞↹⇌⇛↨⇛⇯⇊↾↹⇒←↙⇌⇛↹⇷⇯⇛⇞↟↨⇴↨⇈↹⇠⇌⇛⇯←←↹↷⇠←↙⇛↹↷⇊↹↷⇠←↹⇠↤←⇒⇴⇒↟↹⇷⇯⇴↷↟⇒⇈↝⇛↹↟↹⇷⇛⇒⇙⇞↟↨←↹↳⇴⇌⇠↟↳⇴⇒⇈↝⇊↾↹↲⇴⇒⇒↹⇰⇴↹⇷⇛⇠⇒←↤↝←←↹⇞←↨↷←⇯↨⇛←↹⇰⇴↤⇴↝↟←↹⇌⇙⇯⇠⇴↹↘⇛↨↞↹⇌⇛↝←⇞↝⇛↹↞↹↝↟⇞←↙⇛↹↝←↹⇛↲←⇆⇴⇏'

In [ ]:
counts_text = Counter(text)
print(len(counts_text))

28


In [ ]:
# предположение, что текст на русском т.к. в английском 26 букв
decryption_bigrams = decryption_mcmc_bigrams(text, frequency_bigrams_WAP_rus, counts_WAP, 10000)
decryption_bigrams

'если вы вижите норзальный или помти норзальный текст у чтого соодшения который легко промитать скорее всего вы все сжелали правильно и полумите заксизальный далл ба послежнее метвертое бажание курса хотя конемно я нимего не одешаю'

In [ ]:
text ='←⇠⇒↟↹↷⇊↹↷↟↤↟↨←↹↝⇛⇯↳⇴⇒⇈↝⇊↾↹↟⇒↟↹⇷⇛⇞↨↟↹↝⇛⇯↳⇴⇒⇈↝⇊↾↹↨←⇌⇠↨↹⇙↹⇸↨⇛↙⇛↹⇠⇛⇛↲⇆←↝↟↞↹⇌⇛↨⇛⇯⇊↾↹⇒←↙⇌⇛↹⇷⇯⇛⇞↟↨⇴↨⇈↹⇠⇌⇛⇯←←↹↷⇠←↙⇛↹↷⇊↹↷⇠←↹⇠↤←⇒⇴⇒↟↹⇷⇯⇴↷↟⇒⇈↝⇛↹↟↹⇷⇛⇒⇙⇞↟↨←↹↳⇴⇌⇠↟↳⇴⇒⇈↝⇊↾↹↲⇴⇒⇒↹⇰⇴↹⇷⇛⇠⇒←↤↝←←↹⇞←↨↷←⇯↨⇛←↹⇰⇴↤⇴↝↟←↹⇌⇙⇯⇠⇴↹↘⇛↨↞↹⇌⇛↝←⇞↝⇛↹↞↹↝↟⇞←↙⇛↹↝←↹⇛↲←⇆⇴⇏'
decryption_bigrams = decryption_mcmc_bigrams(text, frequency_bigrams_WAP_rus, counts_WAP, 10000)
decryption_bigrams

'если вы вижите норзальный или помти норзальный текст у чтого соодшения который легко промитать скорее всего вы все сжелали правильно и полумите заксизальный далл ба послежнее метвертое бажание курса хотя конемно я нимего не одешаю'

#5. Бонус: а что если от биграмм перейти к триграммам (тройкам букв) или даже больше?

In [ ]:
def frequency_threegram(text):
  list_a = []
  list_b = []
  for i in range(len(text) - 2):
    if text[i: i+3] not in list_a:
      list_a.append(text[i: i+3])
      list_b.append(text.count(text[i: i+3]) / (len(text) / 3))
  return dict(zip(list_a, list_b))

In [ ]:
frequency_threegram_WAPE_eng = frequency_threegram(' '.join(WarAndPeaceEng_eng))

In [ ]:
len(frequency_threegram_WAPE_eng)

6531

In [ ]:
frequency_threegram_AK_rus = frequency_threegram(' '.join(AnnaKarenina_rus))

In [ ]:
frequency_threegram_WAP_rus = frequency_threegram(' '.join(WarAndPeace_rus))

In [ ]:
def sum_prob(decryption, frequency_bigrams, frequency_threegram):
  sum_dec = 0
  for i in range(0, len(decryption) - 1, 2):
    try:
      sum_dec += frequency_bigrams[decryption[i: i+2]]
    except:
      sum_dec -= min(frequency_bigrams.values())
  
  for i in range(0, len(decryption) - 2, 3):
    try:
      sum_dec += frequency_threegram[decryption[i: i+3]]
    except:
      sum_dec -= min(frequency_threegram.values())
  
  return sum_dec

In [ ]:
def decryption_mcmc_threegram(encrypt, frequency_threegram, frequency_bigrams_ex, frequency_ex, count_iter):
  counts_dec = Counter(encrypt)
  list_a = [i[0] for i in counts_dec.most_common()]
  list_b = [i[0] for i in frequency_ex.most_common(len(list_a))]
  mapping_old = dict(zip(list_a, list_b))
  decryption_old = ''
  for i in encrypt:
    decryption_old += mapping_old[i]

  sum_old = sum_prob(decryption_old, frequency_bigrams_ex, frequency_threegram)
  sum_best = sum_old
  mapping_best = mapping_old.copy()
  decryption_best = decryption_old
  for _ in range(count_iter):
    mapping_new = mapping_old.copy()
    ran_1 = random.choice(list(mapping_new))
    ran_2 = random.choice(list(mapping_new))
    if ran_1 != ran_2:      
      mapping_new[ran_1], mapping_new[ran_2] = mapping_new[ran_2], mapping_new[ran_1]
      decryption_new = ''
      for i in encrypt:
        decryption_new += mapping_new[i]
      sum_new = sum_prob(decryption_new, frequency_bigrams_ex, frequency_threegram)

      if sum_new >= sum_old:
        sum_old = sum_new
        mapping_old = mapping_new.copy()
        if sum_new >= sum_best: 
          sum_best = sum_new
          mapping_best = mapping_new.copy()
      else:
        if sum_new - sum_old > random.random():
          sum_old = sum_new          
          mapping_old = mapping_new.copy()
  decryption_best = ''
  for i in encrypt:
    decryption_best += mapping_best[i]    
  return decryption_best

In [ ]:
# 0.8720136518771331 - качество при биграммах, при добавлении триграмм качество улучшилось 
decryption_AK_threegram_rus = decryption_mcmc_threegram(encrypt_AK_bigrams_rus, frequency_threegram_AK_rus, frequency_bigrams_AK_rus, counts_AK, 100000)
print(decryption_AK_threegram_rus)
print(probability_decryption(decryption_AK_threegram_rus, ' '.join(AnnaKarenina_rus[:4])))

   анна каренина опин из самыу знаменитыу романов льва толстого надинается ставшей ахоризмом хразой все сдастливые семьи чоуоби прщг на прщга кабпая несдастливая семья несдастлива чосвоемщ это книга о ведныу женностяу о люцви о вере о семье о деловедеском постоинстве            лев толстойроман широкого пыуания  дасть червая   лев толстой  анна каренина     роман широкого пыуания     анна каренина чоразила современников всепневностью сопербания неоцыдайная своцопа раскованность човествования щпивительно содетались в этом романе с жельностью ущпобественного взгляпа автора на бизнь
0.909556313993174


In [ ]:
# 0.8720136518771331 - качество при биграммах, при добавлении триграмм качество улучшилось 
decryption_AK_threegram_rus = decryption_mcmc_threegram(encrypt_AK_bigrams_rus, frequency_threegram_AK_rus, {' ': 0}, counts_AK, 100000)
print(decryption_AK_threegram_rus)
print(probability_decryption(decryption_AK_threegram_rus, ' '.join(AnnaKarenina_rus[:4])))

   енне седаныне ичын ыэ оетцх энетанывцх дитенип рупе вировили некынеавой овепгая ежидыэтит ждеэия поа океоврыпца оатуы михибы чдьл не чдьле себчей наокеоврыпей оатуй наокеоврыпе миопиать шви сныле и пакнцх занниовйх и рющпы и пада и оатуа и карипакаосит чиовиыновпа            рап вировиядитен гыдисили чцхеный  кеову мадпей   рап вировия  енне седаныне     дитен гыдисили чцхеный     енне седаныне мидеэыре оипдатаннысип поачнапниовую оичадбеный наищцкеяней опищиче деосипенниову мипаовпипеный ьчыпываруни оикаверыоу п швит дитена о заруниовую хьчибаовпаннили пэлрйче епвиде не быэну
0.26791808873720135


In [ ]:
# 0.8170028818443804 - качество при биграммах, при добавлении триграмм качество улучшилось
decryption_WAP_threegram_rus = decryption_mcmc_threegram(encrypt_WAP_bigrams_rus, frequency_threegram_WAP_rus, frequency_bigrams_WAP_rus, counts_WAP, 10000)
print(decryption_WAP_threegram_rus)
print(probability_decryption(decryption_WAP_threegram_rus, ' '.join(WarAndPeace_rus[:4])))

   нойва и мир  самый изнествый ромав льна виколаенида толстого как викакое жругое произнежевие писателя отрачает глубиву его мироощущевия и эилосоэиицта квига из разряжа недвых потому дто ова обо нсем  о чизви и смерти о любни и дести о мучестне и героизме о слане и пожниге о нойве и миреперный том звакомит с нысшим общестном россии  нека показавы нзаимоотвошевия мечжу рожителями и жетьми н семье ростоных снатонстно у болковских ивтриги у безухоных недера н салове эрейливы апшерер балы н москне и петербурге            лен виколаенид толстойдасть перная  дасть орая  дасть третья              лен виколаенид толстой  нойва и мир  том      дасть перная          е              поместья    
0.8789625360230547


In [ ]:
# 0.8170028818443804 - качество при биграммах, при использовании только триграмм качество улучшилось
decryption_WAP_threegram_rus = decryption_mcmc_threegram(encrypt_WAP_bigrams_rus, frequency_threegram_WAP_rus, {' ': 0}, counts_WAP, 10000)
print(decryption_WAP_threegram_rus)
print(probability_decryption(decryption_WAP_threegram_rus, ' '.join(WarAndPeace_rus[:4])))

   нойла и мир  самый изнестлый ромал вьна ликоваенича товстого как ликакое дрягое произнеделие писатеву отражает гвябиля его мироощящелиу и эивосоэииюта клига из разруда нечлых потомя что ола обо нсем  о жизли и смерти о вцбни и чести о мяжестне и героизме о сване и подниге о нойле и миреперный том злакомит с нысшим общестном россии  нека показалы нзаимоотлошелиу междя родитевуми и детьми н семье ростоных снатонстно я бовколских илтриги я безяхоных нечера н саволе эрейвилы апшерер бавы н москне и петербярге            вен ликоваенич товстойчасть пернау  часть орау  часть третьу              вен ликоваенич товстой  нойла и мир  том      часть пернау          е              поместьу    
0.8515850144092219


In [ ]:
#0.987220447284345 - качество при биграммах, при добавлении триграмм качество ухудшилось
decryption_WAPE_threegram_eng = decryption_mcmc_threegram(encrypt_WAPE_bigrams_eng, frequency_threegram_WAPE_eng, frequency_bigrams_WAPE_eng, counts_WAPE, 10000)
print(decryption_WAPE_threegram_eng)
print(probability_decryption(decryption_WAPE_threegram_eng, ' '.join(WarAndPeaceEng_eng[10:15])))

bhat a kirulent attacv rewlied the wrince not in the least disconcerted py this recewtion he had just entered bearing an emproidered court uniform vnee preeches and shoes and had stars on his preast and a serene exwression on his flat face he swove in that refined french in bhich our grandfathers not only swove put thought and bith the gentle watronizing intonation natural to a man of imwortance bho had grobn old in society and at court he bent uw to anna waklokna vissed her hand wresenting to her his pald scented and shining head and comwlacently seated himself on the sofa  first of all dear friend tell me hob you are
0.9456869009584664


In [ ]:
#0.987220447284345 - качество при биграммах, при использовании только триграмм качество не изменилось
decryption_WAPE_threegram_eng = decryption_mcmc_threegram(encrypt_WAPE_bigrams_eng, frequency_threegram_WAPE_eng, {' ': 0}, counts_WAPE, 10000)
print(decryption_WAPE_threegram_eng)
print(probability_decryption(decryption_WAPE_threegram_eng, ' '.join(WarAndPeaceEng_eng[10:15])))

what a kirulent attacv replied the prince not in the least disconcerted by this reception he had just entered wearing an embroidered court uniform vnee breeches and shoes and had stars on his breast and a serene expression on his flat face he spove in that refined french in which our grandfathers not only spove but thought and with the gentle patronizing intonation natural to a man of importance who had grown old in society and at court he went up to anna paklokna vissed her hand presenting to her his bald scented and shining head and complacently seated himself on the sofa  first of all dear friend tell me how you are
0.987220447284345


### __Качество предсказания букв на методе основанном на MCMC-сэмплировании биграмм и триграмм около 90 процентов__: AnnaKarenina - 0.909556313993174; WarAndPeaceEng(c 10 по 14 предложение) - 0.9456869009584664; WarAndPeace - 0.8789625360230547


### __Качество предсказания букв на методе основанном на MCMC-сэмплировании только триграмм около 80 процентов__: AnnaKarenina - 0.26791808873720135; WarAndPeaceEng(c 10 по 14 предложение) - 0.987220447284345; WarAndPeace - 0.8515850144092219

### __Использование и биграмм и триграмм в MCMC-сэмплировании дает результаты лучше__